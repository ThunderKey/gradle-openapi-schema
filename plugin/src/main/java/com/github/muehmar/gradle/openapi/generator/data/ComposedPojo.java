package com.github.muehmar.gradle.openapi.generator.data;

import ch.bluecare.commons.data.PList;

import java.util.Objects;

/**
 * Represents a composition of other definitions, i.e. pojos. Distinguishes between different types
 * of compositions, see {@link CompositionType}.
 */
public class ComposedPojo {
  private final CompositionType type;
  private final PList<String> pojoNames;
  private final PList<PojoMember> members;

  public enum CompositionType {
    ALL_OF,
    ANY_OF,
    ONE_OF;
  }

  public ComposedPojo(CompositionType type, PList<String> pojoNames, PList<PojoMember> members) {
    this.type = type;
    this.pojoNames = pojoNames;
    this.members = members;
  }

  public CompositionType getType() {
    return type;
  }

  public PList<String> getPojoNames() {
    return pojoNames;
  }

  public PList<PojoMember> getMembers() {
    return members;
  }

  @Override
  public boolean equals(Object o) {

    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    ComposedPojo that = (ComposedPojo) o;
    return type == that.type
        && Objects.equals(pojoNames, that.pojoNames)
        && Objects.equals(members, that.members);
  }

  @Override
  public int hashCode() {
    return Objects.hash(type, pojoNames, members);
  }

  @Override
  public String toString() {
    return "ComposedPojo{"
        + "type="
        + type
        + ", pojoNames="
        + pojoNames
        + ", members="
        + members
        + '}';
  }
}
