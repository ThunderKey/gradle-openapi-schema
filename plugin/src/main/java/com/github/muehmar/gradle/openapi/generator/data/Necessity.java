package com.github.muehmar.gradle.openapi.generator.data;

/** Defines the necessity of an attribute, i.e. if the attribute is required or optional. */
public enum Necessity {
  REQUIRED,
  OPTIONAL;

  public static Necessity fromBoolean(boolean necessity) {
    return necessity ? REQUIRED : OPTIONAL;
  }
}
