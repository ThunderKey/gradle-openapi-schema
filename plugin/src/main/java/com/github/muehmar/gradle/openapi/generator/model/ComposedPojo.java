package com.github.muehmar.gradle.openapi.generator.model;

import ch.bluecare.commons.data.PList;
import lombok.EqualsAndHashCode;
import lombok.ToString;

/**
 * Represents a composition of other definitions, i.e. pojos. Distinguishes between different types
 * of compositions, see {@link CompositionType}.
 */
@EqualsAndHashCode
@ToString
public class ComposedPojo {
  private final PojoName name;
  private final String description;
  private final CompositionType type;
  private final PList<PojoName> pojoNames;
  private final PList<PojoSchema> pojoSchemas;

  public enum CompositionType {
    ALL_OF("AllOf"),
    ANY_OF("AnyOf"),
    ONE_OF("OneOf");

    private final String value;

    CompositionType(String value) {
      this.value = value;
    }

    public String asPascalCaseName() {
      return value;
    }
  }

  public ComposedPojo(
      PojoName name,
      String description,
      CompositionType type,
      PList<PojoName> pojoNames,
      PList<PojoSchema> pojoSchemas) {
    this.name = name;
    this.description = description;
    this.type = type;
    this.pojoNames = pojoNames;
    this.pojoSchemas = pojoSchemas;
  }

  public PojoName getName() {
    return name;
  }

  public String getDescription() {
    return description;
  }

  public String getSuffix() {
    return getName().getSuffix();
  }

  public CompositionType getType() {
    return type;
  }

  public PList<PojoName> getPojoNames() {
    return pojoNames;
  }

  public PList<PojoSchema> getPojoSchemas() {
    return pojoSchemas;
  }
}
