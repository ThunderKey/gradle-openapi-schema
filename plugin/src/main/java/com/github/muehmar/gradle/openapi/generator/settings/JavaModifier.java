package com.github.muehmar.gradle.openapi.generator.settings;

import ch.bluecare.commons.data.PList;
import java.io.Serializable;
import java.util.Optional;

public enum JavaModifier implements Serializable {
  PRIVATE("private"),
  PROTECTED("protected"),
  PACKAGE_PRIVATE("package-private"),
  PUBLIC("public");

  private final String value;

  JavaModifier(String value) {
    this.value = value;
  }

  public String getValue() {
    return value;
  }

  public static Optional<JavaModifier> fromString(String value) {
    return PList.of(values()).find(modifier -> modifier.value.equalsIgnoreCase(value));
  }
}
