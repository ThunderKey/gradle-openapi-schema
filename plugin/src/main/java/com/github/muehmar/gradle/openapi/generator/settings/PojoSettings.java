package com.github.muehmar.gradle.openapi.generator.settings;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.pojoextension.annotations.PojoExtension;
import java.io.Serializable;
import java.util.List;
import java.util.Objects;

@PojoExtension
public class PojoSettings extends PojoSettingsExtension implements Serializable {
  private final JsonSupport jsonSupport;
  private final String packageName;
  private final String suffix;
  private final boolean enableSafeBuilder;
  private final boolean enableConstraints;
  private final List<ClassTypeMapping> classTypeMappings;
  private final List<FormatTypeMapping> formatTypeMappings;
  private final EnumDescriptionSettings enumDescriptionSettings;

  PojoSettings(
      JsonSupport jsonSupport,
      String packageName,
      String suffix,
      boolean enableSafeBuilder,
      boolean enableConstraints,
      List<ClassTypeMapping> classTypeMappings,
      List<FormatTypeMapping> formatTypeMappings,
      EnumDescriptionSettings enumDescriptionSettings) {
    this.jsonSupport = jsonSupport;
    this.packageName = packageName;
    this.suffix = suffix;
    this.enableSafeBuilder = enableSafeBuilder;
    this.enableConstraints = enableConstraints;
    this.classTypeMappings = classTypeMappings;
    this.formatTypeMappings = formatTypeMappings;
    this.enumDescriptionSettings = enumDescriptionSettings;
  }

  public PojoSettings(
      JsonSupport jsonSupport,
      String packageName,
      String suffix,
      boolean enableSafeBuilder,
      boolean enableConstraints,
      PList<ClassTypeMapping> classTypeMappings,
      PList<FormatTypeMapping> formatTypeMappings,
      EnumDescriptionSettings enumDescriptionSettings) {
    this.jsonSupport = jsonSupport;
    this.packageName = packageName;
    this.suffix = suffix;
    this.enableSafeBuilder = enableSafeBuilder;
    this.enableConstraints = enableConstraints;
    this.classTypeMappings = classTypeMappings.toArrayList();
    this.formatTypeMappings = formatTypeMappings.toArrayList();
    this.enumDescriptionSettings = enumDescriptionSettings;
  }

  public String getPackageName() {
    return packageName;
  }

  public String getSuffix() {
    return suffix;
  }

  public JsonSupport getJsonSupport() {
    return jsonSupport;
  }

  public boolean isJacksonJson() {
    return jsonSupport.equals(JsonSupport.JACKSON);
  }

  public boolean isEnableConstraints() {
    return enableConstraints;
  }

  public PList<ClassTypeMapping> getClassTypeMappings() {
    return PList.fromIter(classTypeMappings);
  }

  public PList<FormatTypeMapping> getFormatTypeMappings() {
    return PList.fromIter(formatTypeMappings);
  }

  public boolean isEnableSafeBuilder() {
    return enableSafeBuilder;
  }

  public boolean isDisableSafeBuilder() {
    return !isEnableSafeBuilder();
  }

  public EnumDescriptionSettings getEnumDescriptionSettings() {
    return enumDescriptionSettings;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    PojoSettings that = (PojoSettings) o;
    return enableSafeBuilder == that.enableSafeBuilder
        && enableConstraints == that.enableConstraints
        && jsonSupport == that.jsonSupport
        && Objects.equals(packageName, that.packageName)
        && Objects.equals(suffix, that.suffix)
        && Objects.equals(classTypeMappings, that.classTypeMappings)
        && Objects.equals(formatTypeMappings, that.formatTypeMappings)
        && Objects.equals(enumDescriptionSettings, that.enumDescriptionSettings);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
        jsonSupport,
        packageName,
        suffix,
        enableSafeBuilder,
        enableConstraints,
        classTypeMappings,
        formatTypeMappings,
        enumDescriptionSettings);
  }

  @Override
  public String toString() {
    return "PojoSettings{"
        + "jsonSupport="
        + jsonSupport
        + ", packageName='"
        + packageName
        + '\''
        + ", suffix='"
        + suffix
        + '\''
        + ", enableSafeBuilder="
        + enableSafeBuilder
        + ", enableConstraints="
        + enableConstraints
        + ", classTypeMappings="
        + classTypeMappings
        + ", formatTypeMappings="
        + formatTypeMappings
        + ", enumDescriptionExtraction="
        + enumDescriptionSettings
        + '}';
  }
}
