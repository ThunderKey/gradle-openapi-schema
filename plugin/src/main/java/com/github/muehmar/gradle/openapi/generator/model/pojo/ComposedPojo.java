package com.github.muehmar.gradle.openapi.generator.model.pojo;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.model.Discriminator;
import com.github.muehmar.gradle.openapi.generator.model.Pojo;
import com.github.muehmar.gradle.openapi.generator.model.PojoName;
import com.github.muehmar.gradle.openapi.generator.model.Type;
import com.github.muehmar.gradle.openapi.generator.model.UnresolvedComposedPojo;
import java.util.Optional;
import java.util.function.Function;

public class ComposedPojo implements Pojo {
  private final PojoName name;
  private final String description;
  private final PList<Pojo> pojos;
  private final CompositionType compositionType;
  private final Optional<Discriminator> discriminator;

  private ComposedPojo(
      PojoName name,
      String description,
      PList<Pojo> pojos,
      CompositionType compositionType,
      Optional<Discriminator> discriminator) {
    this.name = name;
    this.description = description;
    this.pojos = pojos;
    this.compositionType = compositionType;
    this.discriminator = discriminator;
  }

  public static ComposedPojo resolvedAnyOf(
      PList<Pojo> pojos, UnresolvedComposedPojo unresolvedComposedPojo) {
    return resolved(pojos, CompositionType.ANY_OF, unresolvedComposedPojo);
  }

  public static ComposedPojo resolvedOneOf(
      PList<Pojo> pojos, UnresolvedComposedPojo unresolvedComposedPojo) {
    return resolved(pojos, CompositionType.ONE_OF, unresolvedComposedPojo);
  }

  private static ComposedPojo resolved(
      PList<Pojo> resolvedPojos,
      CompositionType compositionType,
      UnresolvedComposedPojo unresolvedComposedPojo) {
    return new ComposedPojo(
        unresolvedComposedPojo.getName(),
        unresolvedComposedPojo.getDescription(),
        resolvedPojos,
        compositionType,
        unresolvedComposedPojo.getDiscriminator());
  }

  public enum CompositionType {
    ANY_OF,
    ONE_OF
  }

  @Override
  public PojoName getName() {
    return name;
  }

  @Override
  public String getDescription() {
    return description;
  }

  public PList<Pojo> getPojos() {
    return pojos;
  }

  public CompositionType getCompositionType() {
    return compositionType;
  }

  public Optional<Discriminator> getDiscriminator() {
    return discriminator;
  }

  @Override
  public Pojo addObjectTypeDescription(PojoName objectTypeName, String description) {
    return new ComposedPojo(
        name,
        description,
        pojos.map(pojo -> pojo.addObjectTypeDescription(objectTypeName, description)),
        compositionType,
        discriminator);
  }

  @Override
  public Pojo inlineObjectReference(
      PojoName referenceName, String referenceDescription, Type referenceType) {
    return new ComposedPojo(
        name,
        description,
        pojos.map(
            pojo -> pojo.inlineObjectReference(referenceName, referenceDescription, referenceType)),
        compositionType,
        discriminator);
  }

  @Override
  public <T> T fold(
      Function<ObjectPojo, T> onObjectPojo,
      Function<ArrayPojo, T> onArrayType,
      Function<EnumPojo, T> onEnumPojo,
      Function<ComposedPojo, T> onComposedPojo,
      Function<FreeFormPojo, T> onFreeFormPojo) {
    return onComposedPojo.apply(this);
  }
}
