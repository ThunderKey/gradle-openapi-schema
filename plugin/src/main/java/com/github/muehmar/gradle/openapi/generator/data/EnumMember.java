package com.github.muehmar.gradle.openapi.generator.data;

import ch.bluecare.commons.data.PList;
import ch.bluecare.commons.data.Pair;
import com.github.muehmar.gradle.openapi.generator.settings.EnumDescriptionSettings;
import java.util.Objects;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/** An enum member containing the name itself and optionally a description. */
public class EnumMember {
  private final Name name;
  private final String description;

  public EnumMember(Name name, String description) {
    this.name = name;
    this.description = description;
  }

  /** */
  public static Optional<EnumMember> extractDescription(
      Name memberName, EnumDescriptionSettings settings, String description) {
    if (settings.isDisabled()) {
      return Optional.empty();
    }

    final String prefixMatcher = settings.getPrefixMatcherForMember(memberName);
    final Pattern pattern = Pattern.compile(".*" + prefixMatcher + "(.*)");

    return PList.fromArray(description.split("\n"))
        .map(pattern::matcher)
        .filter(Matcher::find)
        .map(matcher -> matcher.group(1))
        .headOption()
        .map(desc -> new EnumMember(memberName, desc.trim()));
  }

  /**
   * Returns a list of enums with found description. If no description is found, the description
   * will be set to an empty string in case failOnMissingDescription is turned off. In case it is
   * turned on, this method will fail if a description of a member is not found.
   */
  public static PList<EnumMember> extractDescriptions(
      PList<Name> enumMembers, EnumDescriptionSettings settings, String description) {
    if (settings.isDisabled()) {
      return enumMembers.map(name -> new EnumMember(name, ""));
    }

    final PList<Optional<EnumMember>> enumWithDescriptions =
        enumMembers.map(m -> extractDescription(m, settings, description));

    int countOfMembersWithDescription = enumWithDescriptions.flatMapOptional(e -> e).size();
    final boolean incompleteDescription =
        countOfMembersWithDescription != enumMembers.size() && countOfMembersWithDescription != 0;

    if (settings.isFailOnIncompleteDescriptions() && incompleteDescription) {
      final String missingEnumMembers =
          enumWithDescriptions
              .zip(enumMembers)
              .filter(p -> !p.first().isPresent())
              .map(Pair::second)
              .map(Name::asString)
              .mkString(", ");

      final String message =
          String.format(
              "No description found for the following enum members [%s]. "
                  + "Either complete the description for these members or turn off the "
                  + "failOnMissingDescription setting for the enum extraction.",
              missingEnumMembers);
      throw new IllegalStateException(message);
    } else {
      return enumWithDescriptions
          .zip(enumMembers)
          .map(p -> p.first().orElse(new EnumMember(p.second(), "")));
    }
  }

  public Name getName() {
    return name;
  }

  public String getDescription() {
    return description;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    EnumMember that = (EnumMember) o;
    return Objects.equals(name, that.name) && Objects.equals(description, that.description);
  }

  @Override
  public int hashCode() {
    return Objects.hash(name, description);
  }

  @Override
  public String toString() {
    return "EnumMember{" + "name=" + name + ", description='" + description + '\'' + '}';
  }
}
