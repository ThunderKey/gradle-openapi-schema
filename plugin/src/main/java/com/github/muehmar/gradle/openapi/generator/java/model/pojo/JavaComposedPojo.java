package com.github.muehmar.gradle.openapi.generator.java.model.pojo;

import static com.github.muehmar.gradle.openapi.generator.model.pojo.ComposedPojo.CompositionType.ANY_OF;
import static com.github.muehmar.gradle.openapi.generator.model.pojo.ComposedPojo.CompositionType.ONE_OF;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.model.Discriminator;
import com.github.muehmar.gradle.openapi.generator.model.PojoName;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Constraints;
import com.github.muehmar.gradle.openapi.generator.model.pojo.ComposedPojo;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import java.util.Optional;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class JavaComposedPojo implements JavaPojo {
  private final PojoName name;
  private final String description;
  private final PList<JavaPojo> javaPojos;
  private final ComposedPojo.CompositionType compositionType;
  private final Constraints constraints;
  private final Optional<Discriminator> discriminator;

  JavaComposedPojo(
      PojoName name,
      String description,
      PList<JavaPojo> javaPojos,
      ComposedPojo.CompositionType compositionType,
      Constraints constraints,
      Optional<Discriminator> discriminator) {
    this.name = name;
    this.description = Optional.ofNullable(description).orElse("");
    this.javaPojos = javaPojos;
    this.compositionType = compositionType;
    this.constraints = constraints;
    this.discriminator = discriminator;
  }

  public static JavaComposedPojo wrap(ComposedPojo composedPojo, TypeMappings typeMappings) {
    return new JavaComposedPojo(
        composedPojo.getName(),
        composedPojo.getDescription(),
        composedPojo.getPojos().map(pojo -> JavaPojo.wrap(pojo, typeMappings)),
        composedPojo.getCompositionType(),
        composedPojo.getConstraints(),
        composedPojo.getDiscriminator());
  }

  @Override
  public PojoName getName() {
    return name;
  }

  @Override
  public String getDescription() {
    return description;
  }

  public PList<JavaPojo> getJavaPojos() {
    return javaPojos;
  }

  public ComposedPojo.CompositionType getCompositionType() {
    return compositionType;
  }

  public boolean isAnyOf() {
    return compositionType.equals(ANY_OF);
  }

  public boolean isOneOf() {
    return compositionType.equals(ONE_OF);
  }

  public Optional<Discriminator> getDiscriminator() {
    return discriminator;
  }

  public JavaObjectPojo wrapIntoJavaObjectPojo() {
    return JavaObjectPojo.from(name, description, getMembers(), constraints);
  }

  public PList<JavaPojoMember> getMembers() {
    return javaPojos.flatMap(JavaPojo::getMembersOrEmpty).distinct(Function.identity());
  }

  @Override
  public <T> T fold(
      Function<JavaArrayPojo, T> onArrayPojo,
      Function<JavaEnumPojo, T> onEnumPojo,
      Function<JavaObjectPojo, T> onObjectPojo,
      Function<JavaComposedPojo, T> onComposedPojo,
      Function<JavaFreeFormPojo, T> onFreeFormPojo) {
    return onComposedPojo.apply(this);
  }
}
