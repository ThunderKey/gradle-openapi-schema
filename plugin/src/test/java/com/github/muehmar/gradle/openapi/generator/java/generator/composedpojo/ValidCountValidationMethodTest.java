package com.github.muehmar.gradle.openapi.generator.java.generator.composedpojo;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaComposedPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos;
import com.github.muehmar.gradle.openapi.generator.model.pojo.ComposedPojo;
import com.github.muehmar.gradle.openapi.generator.settings.JavaModifier;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import java.util.stream.Stream;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

@ExtendWith(SnapshotExtension.class)
class ValidCountValidationMethodTest {
  private Expect expect;

  public static Stream<Arguments> compositionTypeAndSettings() {
    final PList<ComposedPojo.CompositionType> compositionTypes =
        PList.of(ComposedPojo.CompositionType.values());
    final PojoSettings validationMethodsSettings =
        TestPojoSettings.defaultSettings()
            .withValidationMethods(
                TestPojoSettings.defaultValidationMethods()
                    .withModifier(JavaModifier.PROTECTED)
                    .withDeprecatedAnnotation(true));
    final PList<PojoSettings> settings =
        TestPojoSettings.validationVariants().add(validationMethodsSettings);
    return compositionTypes
        .flatMap(type -> settings.map(setting -> Arguments.arguments(type, setting)))
        .toStream();
  }

  @ParameterizedTest
  @MethodSource("compositionTypeAndSettings")
  @SnapshotName("ValidCountValidationMethod")
  void generate_when_pojoAndSettings_then_correctOutput(
      ComposedPojo.CompositionType type, PojoSettings settings) {
    final Generator<JavaComposedPojo, PojoSettings> generator =
        ValidCountValidationMethod.generator();

    final JavaComposedPojo javaPojo = JavaPojos.composedPojo(type);
    final Writer writer = generator.generate(javaPojo, settings, Writer.createDefault());

    final String scenario =
        PList.of(
                type.name(),
                Boolean.toString(settings.isEnableValidation()),
                settings.getValidationApi().getValue(),
                settings.getValidationMethods().getModifier())
            .mkString(",");

    expect
        .scenario(scenario)
        .toMatchSnapshot(writer.getRefs().mkString("\n") + "\n\n" + writer.asString());
  }
}
