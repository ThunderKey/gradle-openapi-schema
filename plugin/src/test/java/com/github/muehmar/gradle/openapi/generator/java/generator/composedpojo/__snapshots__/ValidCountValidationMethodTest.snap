ValidCountValidationMethod[ANY_OF,false,jakarta-2,PRIVATE]=[
.
.

]


ValidCountValidationMethod[ANY_OF,true,jakarta-2,PRIVATE]=[
javax.validation.constraints.AssertFalse

@AssertFalse(message = "Is not valid against one of the schemas [User, NecessityAndNullability]")
private boolean isValidAgainstNoSchema() {
  return getValidCount() == 0;
}
]


ValidCountValidationMethod[ANY_OF,true,jakarta-2,PROTECTED]=[
javax.validation.constraints.AssertFalse

@AssertFalse(message = "Is not valid against one of the schemas [User, NecessityAndNullability]")
@Deprecated
protected boolean isValidAgainstNoSchema() {
  return getValidCount() == 0;
}
]


ValidCountValidationMethod[ANY_OF,true,jakarta-3,PRIVATE]=[
jakarta.validation.constraints.AssertFalse

@AssertFalse(message = "Is not valid against one of the schemas [User, NecessityAndNullability]")
private boolean isValidAgainstNoSchema() {
  return getValidCount() == 0;
}
]


ValidCountValidationMethod[ONE_OF,false,jakarta-2,PRIVATE]=[
.
.

]


ValidCountValidationMethod[ONE_OF,true,jakarta-2,PRIVATE]=[
javax.validation.constraints.AssertFalse
javax.validation.constraints.AssertFalse

@AssertFalse(message = "Is not valid against one of the schemas [User, NecessityAndNullability]")
private boolean isValidAgainstNoSchema() {
  return getValidCount() == 0;
}

@AssertFalse(message = "Is valid against more than one of the schemas [User, NecessityAndNullability]")
private boolean isValidAgainstMoreThanOneSchema() {
  return getValidCount() > 1;
}
]


ValidCountValidationMethod[ONE_OF,true,jakarta-2,PROTECTED]=[
javax.validation.constraints.AssertFalse
javax.validation.constraints.AssertFalse

@AssertFalse(message = "Is not valid against one of the schemas [User, NecessityAndNullability]")
@Deprecated
protected boolean isValidAgainstNoSchema() {
  return getValidCount() == 0;
}

@AssertFalse(message = "Is valid against more than one of the schemas [User, NecessityAndNullability]")
@Deprecated
protected boolean isValidAgainstMoreThanOneSchema() {
  return getValidCount() > 1;
}
]


ValidCountValidationMethod[ONE_OF,true,jakarta-3,PRIVATE]=[
jakarta.validation.constraints.AssertFalse
jakarta.validation.constraints.AssertFalse

@AssertFalse(message = "Is not valid against one of the schemas [User, NecessityAndNullability]")
private boolean isValidAgainstNoSchema() {
  return getValidCount() == 0;
}

@AssertFalse(message = "Is valid against more than one of the schemas [User, NecessityAndNullability]")
private boolean isValidAgainstMoreThanOneSchema() {
  return getValidCount() > 1;
}
]